backend/local/clone_darwin.go

```diff
@@ -44,11 +50,21 @@ func (f *Fs) Copy(ctx context.Context, src fs.Object, remote string) (fs.Object,
 		return nil, err
 	}
 
-	err = Clone(srcObj.path, f.localPath(remote))
+	srcPath := srcObj.path
+	if f.opt.FollowSymlinks { // in --copy-links mode, find the real file being pointed to and pass that in instead
+		linkPath, err := os.Readlink(srcPath)
```

Comment:
This won't work if the symlink points to another symlink...

Resolving symlinks is a pain as you have to detect loops also, so we should try to get the kernel to do it if we can.

I would have thought we want to run the Clone without `CLONE_NOFOLLOW`/`COPYFILE_NOFOLLOW` or something like that? Though you don't appear to be using those flags so there is a gap in my understanding!

Actually I found [filepath.EvalSymlinks](https://pkg.go.dev/path/filepath?utm_source=godoc#EvalSymlinks) which looks like it would do the job if we can't get Clone to do it

Here is [the source for the meat of filpath.EvalSymlinks](https://cs.opensource.google/go/go/+/refs/tags/go1.23.1:src/path/filepath/symlink.go;l=16) - as you see it is quite non trivial resolving symlinks!

Reply:
> I would have thought we want to run the Clone without `CLONE_NOFOLLOW`/`COPYFILE_NOFOLLOW` or something like that?

This was my first instinct too, and there is a first draft [here](https://github.com/nielash/rclone/commit/736c31803b8dcb6b8ac7952fa45c811447d44098) where I tried that. The problem is... while it does fix the bug, it seems to just copy instead of clone.

It is possible that there is some combination of flags that would do what we want, but that one wasn't it. Any ideas about what it might be?

I also came across [this comment](https://github.com/apple-oss-distributions/copyfile/blob/ed3f0a8bf8b6bac6838c92c297afcc826fec75f4/copyfile.c#L736) which seems to suggest it might not be possible...

Reply:
That comment in the kernel looks very relevant

    // or we're cloning, where we never follow symlinks.

Do you want to update with `filepath.EvalSymlinks` then instead of `os.Readlink` - that should be a robust solution if Clone can't do it for us?

Reply:
Yes, that sounds right to me.

Reply:
Made that change.